src/Npgsql/TypeHandlers/ArrayHandler.cs

```diff
@@ -238,7 +247,7 @@ int ValidateAndGetLengthNonGeneric(ICollection value, ref NpgsqlLengthCache leng
             return len;
         }
 
-        internal override Task WriteWithLengthInternal<TAny>(TAny value, NpgsqlWriteBuffer buf, NpgsqlLengthCache lengthCache, NpgsqlParameter parameter, bool async)
+        internal override Task WriteWithLengthInternal<TAny>(TAny value, NpgsqlWriteBuffer buf, NpgsqlLengthCache? lengthCache, NpgsqlParameter? parameter, bool async)
```

Comment:
This method could possibly benefit from refactoring.
We can save the bang, a duplicate check and some redundancy.
My suggestion:
```csharp
internal override Task WriteWithLengthInternal<TAny>(TAny value, NpgsqlWriteBuffer buf, NpgsqlLengthCache? lengthCache, NpgsqlParameter? parameter, bool async)
{
    if (buf.WriteSpaceLeft < 4)
        return WriteWithLengthLong();

    return WriteWithLength();

    async Task WriteWithLengthLong()
    {
        await buf.Flush(async);
        await WriteWithLength();
    }

    Task WriteWithLength()
    {
        if (value == null || typeof(TAny) == typeof(DBNull))
        {
            buf.WriteInt32(-1);
            return Task.CompletedTask;
        }

        buf.WriteInt32(ValidateAndGetLength(value, ref lengthCache, parameter));

        if (value is ICollection<TElement> list)
            return WriteGeneric(list, buf, lengthCache, async);

        if (value is ICollection nonGeneric)
            return WriteNonGeneric(nonGeneric, buf, lengthCache, async);

        throw CantWriteTypeException(value.GetType());
    }
}
```

Reply:
Yep, that looks better! Made the change in its own commit.

(although there's no bang being saved here...)

Reply:
> (although there's no bang being saved here...)

https://github.com/npgsql/npgsql/pull/2450/commits/4e57cd128a2aacc5a9bc5ec4a852b173134f2ba1#diff-2a98bdfdf7ed91336835a08e0fc42264

```csharp
throw CantWriteTypeException(value   !   .GetType());
```
😉